src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,390 @@
+            # Regex '[abc0-9]' (set of characters)
+            negate = value[0][0] == sre.NEGATE
+            if isinstance(pattern, text_type):
+                builder = CharactersBuilder(negate, context.flags)
+            else:
+                builder = BytesBuilder(negate, context.flags)
+
+            for charset_code, charset_value in value:
+                if charset_code == sre.NEGATE:
+                    # Regex '[^...]' (negation)
+                    # handled by builder = CharactersBuilder(...) above
+                    pass
+                elif charset_code == sre.LITERAL:
+                    # Regex '[a]' (single char)
+                    builder.add_char(charset_value)
+                elif charset_code == sre.RANGE:
+                    # Regex '[a-z]' (char range)
+                    low, high = charset_value
+                    for char_code in hrange(low, high + 1):
+                        builder.add_char(char_code)
+                elif charset_code == sre.CATEGORY:
+                    # Regex '[\w]' (char category)
+                    builder.add_category(charset_value)
+                else:  # pragma: no cover
+                    # Currently there are no known code points other than
+                    # handled here. This code is just future proofing
+                    raise AssertionError('Unknown charset code: %s'
+                                         % charset_code)
+            return builder.strategy
+
+        elif code == sre.ANY:
+            # Regex '.' (any char)
+            if isinstance(pattern, text_type):
+                if context.flags & re.DOTALL:
+                    return st.characters()
+                return st.characters(blacklist_characters=u'\n')
+            else:
+                if context.flags & re.DOTALL:
+                    return binary_char
+                return binary_char.filter(lambda c: c != b'\n')
+
+        elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            # An empty string (or newline) will match the token itself, but
+            # we don't and can't check the position (eg '%' at the end)
+            if value == sre.AT_END:
+                newline = u'\n' if isinstance(pattern, text_type) else b'\n'
+                return st.sampled_from([empty, newline])
+            return st.just(empty)
+
+        elif code == sre.SUBPATTERN:
+            # Various groups: '(...)', '(:...)' or '(?P<name>...)'
+            old_flags = context.flags
+            if HAS_SUBPATTERN_FLAGS:  # pragma: no cover
+                # This feature is available only in specific Python versions
+                context.flags = (context.flags | value[1]) & ~value[2]
+
+            strat = _strategy(value[-1], context, pattern)
+
+            context.flags = old_flags
+
+            if value[0]:
+                context.groups[value[0]] = strat
+                strat = st.shared(strat, key=(pattern, value[0]))
```

Comment:
I admit I don't understand how shared shares things (another docs issue!), but that's not what happens here.  `s` is a strategy composed of various things, including a `shared(...)` with the same key.  Both parts of the string are therefore the same.

With identical strategies you get `('foofoo', 'foofoo')`, but that's still a correct output - filtering isn't the problem.  If we explicitly want per-draw-only group sharing (which I agree is nice to have), it's going to be a nastier and more stateful strategy.  

Would you object to merging this as a working-if-imperfect strategy, and opening an issue to improve the niggly bits later?